gapic/src/main/com/google/gapid/views/VirtualTreeViewer.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (C) 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.gapid.views;
+
+import org.eclipse.jface.viewers.IBaseLabelProvider;
+import org.eclipse.jface.viewers.IContentProvider;
+import org.eclipse.jface.viewers.TreeViewer;
+import org.eclipse.swt.graphics.Rectangle;
+import org.eclipse.swt.widgets.Tree;
+import org.eclipse.swt.widgets.TreeItem;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * A {@link TreeViewer} that notifies the bound {@link IContentProvider} and
+ * {@link IBaseLabelProvider} of time visibility changes if they also implement
+ * the {@link VisibilityListener} interface.
+ */
+public class VirtualTreeViewer extends TreeViewer {
+  /**
+   * The interface the {@link IContentProvider} and {@link IBaseLabelProvider} can
+   * implement to be notified about {@link TreeItem}s being made visible and hidden
+   * usually by scrolling.
+   */
+  public interface VisibilityListener {
+    /**
+     * Called when the {@link TreeItem} is made visible.
+     */
+    void OnShow(TreeItem item);
```

Comment:
I know I'm going against the grain here, so I'm completely OK if you don't want to follow my pattern. Below is my opinion, but feel free to completely ignore it without any guilt and I won't bother you about it ever again.

I don't like implicit things. While I understand that you can argue that they are redundant, I would point out that they are only redundant due to the context. When I'm reading code I'd like to spend my "context awareness" brain power on higher level concepts. This is compounded by the fact that typically the absence of a visibility modifier means that it's package protected... except in an interface where it means public. Why force my brain to make that distinction? This is even worse for things like "String foo = ..." In a class context, this means a package protected instance field. In interface this is a "public static final" constant. This is the kind of verbosity I like: you have to type a little bit more, but every time I read it, I have to spend less effort deciphering.